### PR TITLE
feat: add support for soft-deleting subsidy access policies

### PR DESCRIPTION
**Pull request by pwnage101**
ENT-7150

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -68,10 +68,10 @@ coverage: clean
 	$(BROWSER) htmlcov/index.html
 
 isort_check: ## check that isort has been run
-	isort --check-only -rc enterprise_access/
+	isort --check-only enterprise_access/
 
 isort: ## run isort to sort imports in all Python files
-	isort --recursive --atomic enterprise_access/
```
Comment (iloveagent57) [inline Makefile:74]: why no recursive?

File: enterprise_access/apps/core/constants.py [hunk 2]
```diff
@@ -11,10 +11,11 @@
...
 
-POLICY_ADMIN_ROLE = 'enterprise_access_subsidy_access_policy_admin'
-POLICY_LEARNER_ROLE = 'enterprise_access_subsidy_access_policy_learner'
-POLICY_READ_PERMISSION = 'policy.has_read_access'
-POLICY_REDEMPTION_PERMISSION = 'policy.has_redemption_access'
+SUBSIDY_ACCESS_POLICY_OPERATOR_ROLE = 'enterprise_access_subsidy_access_policy_operator'
+SUBSIDY_ACCESS_POLICY_LEARNER_ROLE = 'enterprise_access_subsidy_access_policy_learner'
+SUBSIDY_ACCESS_POLICY_READ_PERMISSION = 'subsidy_access_policy.has_read_access'
+SUBSIDY_ACCESS_POLICY_WRITE_PERMISSION = 'subsidy_access_policy.has_write_access'
+SUBSIDY_ACCESS_POLICY_REDEMPTION_PERMISSION = 'subsidy_access_policy.has_redemption_access'
```
Comment (iloveagent57) [inline enterprise_access/apps/core/constants.py:18]: Could we get this permissions stuff carved out into a different PR/commit, to help isolate for safer rollback?

File: enterprise_access/apps/api/v1/views/subsidy_access_policy.py [hunk 3]
```diff
@@ -117,21 +126,53 @@ def list(self, request, *args, **kwargs):
...
+
+        try:
+            policy_to_soft_delete = self.get_queryset().get(uuid=uuid)
+        except SubsidyAccessPolicy.DoesNotExist:
+            return Response(None, status=status.HTTP_404_NOT_FOUND)
+
+        if policy_to_soft_delete.active:
+            policy_to_soft_delete.active = False
+            policy_to_soft_delete._change_reason = delete_reason  # pylint: disable=protected-access
+            policy_to_soft_delete.save()
```
Comment (iloveagent57) [inline enterprise_access/apps/api/v1/views/subsidy_access_policy.py:156]: nit: might be nice to override the `delete()` method of the model itself to do this.  That way, it's more reusable for any other places we might need to delete policies from.

File: Makefile (hunk 1 above)
Comment (pwnage101) [reply Makefile:74]: --recursive is a deprecated and ignored flag, and now is the default.

File: enterprise_access/settings/base.py [hunk 4]
```diff
@@ -55,6 +55,7 @@ def root(*path_fragments):
     'rest_framework',
     'rest_framework_swagger',
     'rules.apps.AutodiscoverRulesConfig',
+    'simple_history',
```
Comment (pwnage101) [inline enterprise_access/settings/base.py:58]: Interestingly, simple_history middleware was installed, but not simple_history itself.

File: enterprise_access/apps/api/serializers/subsidy_access_policy.py [hunk 5]
```diff
@@ -224,6 +224,19 @@ class SubsidyAccessPolicyCanRedeemRequestSerializer(serializers.Serializer):
...
 
 
+# pylint: disable=abstract-method
+class SubsidyAccessPolicyDeleteRequestSerializer(serializers.Serializer):
+    """
+    Request Serializer for DELETE parameters to an API call to delete a subsidy access policy.
+
+    For view: SubsidyAccessPolicyViewSet.destroy
+    """
+    reason = serializers.CharField(
```
Comment (iloveagent57) [inline enterprise_access/apps/api/serializers/subsidy_access_policy.py:234]: nit: might be worth adding `allow_null=True` here. https://www.django-rest-framework.org/api-guide/fields/#allow_null

File: enterprise_access/apps/subsidy_access_policy/models.py [hunk 6]
```diff
@@ -529,6 +529,15 @@ def resolve_policy(cls, redeemable_policies):
         )
         return sorted_policies[0]
 
+    def delete(self, *args, **kwargs):
+        """
+        Perform a soft-delete, overriding the standard delete() method to prevent hard-deletes.
+
+        If this instance was already soft-deleted, invoking delete() is a no-op.
+        """
+        self.active = False
```
Comment (iloveagent57) [inline enterprise_access/apps/subsidy_access_policy/models.py:538]: nit: you could set the reason in here from kwargs if you wanted.

**Review by iloveagent57 — APPROVED** [review]

File: enterprise_access/apps/api/serializers/subsidy_access_policy.py (hunk 5 above)
Comment (pwnage101) [reply enterprise_access/apps/api/serializers/subsidy_access_policy.py:234]: Oh, i totally forgot to write unit tests for the reason field.

File: enterprise_access/apps/subsidy_access_policy/models.py (hunk 6 above)
Comment (pwnage101) [reply enterprise_access/apps/subsidy_access_policy/models.py:538]: oh nice, I wasn't sure if changing the call signature of delete was kosher

**Pull request merged**